### ISSUE TITLE
secretmanager.Secret requires replace when updating from v6.62.0 to v7.4.0

### ISSUE THREAD

**Issue by lucasknode** [pulumi/pulumi-gcp#1488]
### What happened?

I just upgraded our pulumi-gcp version from 6.62.0 to 7.4.0 which resulted in finding a backwards incompatible change to `pulumi_gcp.secretmanager.SecretReplicationArgs` - The `automatic` arg, which took a bool, has been removed in favor of the `auto` arg which takes an instance of `pulumi_gcp.secretmanager.SecretReplicationAutoArgs`.

When making the required update and trying to update our stack with `pulumi up` it attempts to do a replace of all of our secrets (140 resources). All of these fail with an error like the following:
``` shell
Error creating Secret: googleapi: Error 409: Secret [projects/xxxxxx/secrets/MY_SECRET] already exists.
```

This leads me to 2 issues:
1. I would have expected that changing to use the new `auto` arg would still result in the same underlying resource definition and therefore wouldn't trigger any update on the GCP side
2. If it is necessary, I would think the replace would be able to handle deleting the secret first and then recreating it with the same name (although I suppose it makes sense that it doesn't since this would result in downtime)

Is there some way to resolve this issue?

### Example

**Secret definition using pulumi_gcp  6.62.0:**
```python
secret = gcp.secretmanager.Secret(
    "my-secret",
    project="my-project",
    secret_id="my-secret-id",
    replication=gcp.secretmanager.SecretReplicationArgs(
        automatic=True
    )
)
```

**Secret definition using pulumi_gcp  7.4.0:**
```python
secret = gcp.secretmanager.Secret(
    "my-secret",
    project="my-project",
    secret_id="my-secret-id",
    replication=gcp.secretmanager.SecretReplicationArgs(
        auto=gcp.secretmanager.SecretReplicationAutoArgs()
    )
)
```

### Output of `pulumi about`

CLI          
Version      3.78.1
Go Version   go1.20.7
Go Compiler  gc

Plugins
NAME        VERSION
gcp         7.4.0
kubernetes  4.1.0
postgresql  3.9.0
python      3.10.11
random      4.13.2

Host     
OS       darwin
Version  14.1.2
Arch     arm64

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by iwahbe** [comment]
Hey @lucasknode. I'm sorry you're hitting this issue.

> 1. I would have expected that changing to use the new auto arg would still result in the same underlying resource definition and therefore wouldn't trigger any update on the GCP side

It seems like that is the correct behavior. I'm going to leave this issue open to track that.

> 2. If it is necessary, I would think the replace would be able to handle deleting the secret first and then recreating it with the same name (although I suppose it makes sense that it doesn't since this would result in downtime)

By default, Pulumi is designed to prevent downtime. Each resource that needs to be deleted *before* it is re-created needs to explicitly opt in to that behavior. You can opt in these resources by setting the resource option [`delete_before_replace=True`](https://github.com/pulumi/pulumi/blob/644fa55ef5dee5b45434a1c1a350c1f533dd9bcf/sdk/python/lib/pulumi/resource.py#L275-L278).

**Comment by thovden** [comment]
I'm also hitting this problem. I don't think re-create is a good option here since we already have secret versions that we depend on.

**Comment by thovden** [comment]
@iwahbe does https://github.com/pulumi/pulumi-gcp/commit/3b51797921e137603b2e2ccf9dda1308466b92b7 mean that Puliumi will delete the secrets and recreate them by default? If so I think this is a very dangerous behavior. The secret values we have are not managed by pulumi since they are, well, secrets, and manually added to the the secret once the GCP resource is available. So deleting the secrets and recreating them (without the values) will take down our system. 

So the existing behavior is much better - i.e. fail on deploy. The best behavior, of course, is to keep the underlying GCP secret resoufce while updating the provider. Is that possible?

**Comment by iwahbe** [comment]
> @iwahbe does https://github.com/pulumi/pulumi-gcp/commit/3b51797921e137603b2e2ccf9dda1308466b92b7 mean that Puliumi will delete the secrets and recreate them by default? If so I think this is a very dangerous behavior.

It should, yes. Replace is inherently dangerous with stateful resources. I think that `pulumi_gcp.secretmanager.Secret` should behave like other resources. If you want to make sure that a resource is not deleted, you can [protect](https://www.pulumi.com/docs/concepts/options/protect/) the resource. That will prevent unwanted "replace" and "delete" operations. That protection will cover all deletes, not just those that come from a replace.


> The best behavior, of course, is to keep the underlying GCP secret resoufce while updating the provider. Is that possible?

Yes. That is the goal here.

**Comment by michael-altis** [comment]
I'm encountering the same issue with my Secret resources, and wanted to add that it seems the `gcp:serviceAccount/account:Account` resource is similarly affected, i.e. pulumi wants to recreate it while only the provider field has changed. 

I apparently have various resources created with different versions, so while downgrading has fixed the secret replacement for now, I still cannot perform an `up` right now as it would replace all my service accounts, which would have a similarly undesirable impact. 

Looking forward to a fix, will follow the issue to stay updated.

Versions in case it's useful:
 - before issue: `pulumi==3.73.0`, `pulumi_gcp==6.59.0`
 - current: `pulumi==3.101.0`, `pulumi_gcp==7.5.0`

**Comment by iwahbe** [comment]
Thanks for that @michael-altis. It is helpful.

**Comment by lucasknode** [comment]
@michael-altis we ran into the same issue. Because of that we opted against rolling back since it would have been even more negatively impactful replacing our service accounts. What we ended up doing was staying with Pulumi 7.4.0, and because all of our secrets are managed by Pulumi we did the following:
1. Delete all of our secrets directly via gcloud CLI (with a bash script to loop through them)
3. Run `pulumi up --refresh` so that pulumi would perform the refresh first and detect that the secrets were missing and then go ahead and create them cleanly with the new provider

This wasn't ideal since obviously deleting all the secrets like that incurred downtime for our system, but given the automated way we scripted it all it was ~2mins of downtime. Not ideal, but it unblocked us given that we were completely unable to deploy new infra due to this issue.

Looking forward to the real solution being worked on so we never have to deal with that song and dance again.

**Comment by michael-altis** [comment]
@lucasknode Thanks for the info, indeed, replacing service accounts would incur even more costs due to having to replace the service account json files, replacing secrets is definitely the lesser of two evils here. 

Secret replacement should technically be a safe operation for my setup as the only consumer is Kubernetes External Secrets, and it should keep existing K8s secrets in place if it can't find the remote version. However as I have no important updates for the moment I prefer to wait until there's a fix or I have new infra to deploy.

Out of curiosity, any particular reason to delete with a bash script rather than let Pulumi do the recreate?

**Comment by t0yv0** [comment]
Hi folks, root-caused it to https://github.com/pulumi/pulumi-terraform-bridge/issues/1667 but unable to land that quickly I'm still looking for some Pulumi-level workaround here to make these upgrades uneventful for users. Thanks for your patience.

**Comment by alexcouret** [comment]
This also seems to happen for `RouterNat`. Upgrading from 6.50.0 to 7.38.0 without changing anything requires a replace because of the provider update.
We allocate IPs manually and they're used to whitelist external services so we can't replace the resource. This blocks us from creating a resource using Pulumi that's been introduced in 7.x.

**Comment by VenelinMartinov** [comment]
Hi @alexcouret can you please open a new issue with details of the problem you are facing?
